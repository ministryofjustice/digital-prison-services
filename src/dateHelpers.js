const moment = require('moment')

const DATE_TIME_FORMAT_SPEC = 'YYYY-MM-DDTHH:mm:ss'
const DATE_ONLY_FORMAT_SPEC = 'YYYY-MM-DD'
const DAY_MONTH_YEAR = 'DD/MM/YYYY'
const MOMENT_DAY_OF_THE_WEEK = 'dddd'
const MOMENT_TIME = 'HH:mm'

const DayOfTheWeek = dateTime => moment(dateTime, DATE_TIME_FORMAT_SPEC).format(MOMENT_DAY_OF_THE_WEEK)
const DayMonthYear = dateTime => moment(dateTime, DATE_TIME_FORMAT_SPEC).format(DAY_MONTH_YEAR)
const Time = dateTime => moment(dateTime, DATE_TIME_FORMAT_SPEC).format(MOMENT_TIME)

const buildDateTime = ({ date, hours, minutes, dateFormat = DAY_MONTH_YEAR }) => {
  const time =
    date &&
    Number.isSafeInteger(Number.parseInt(hours, 10)) &&
    Number.isSafeInteger(Number.parseInt(minutes, 10)) &&
    moment(date, dateFormat)
  return time ? time.hour(Number(hours)).minutes(Number(minutes)) : ''
}

module.exports = {
  DATE_TIME_FORMAT_SPEC,
  DATE_ONLY_FORMAT_SPEC,
  DAY_MONTH_YEAR,
  MOMENT_DAY_OF_THE_WEEK,
  MOMENT_TIME,
  DayOfTheWeek,
  DayMonthYear,
  Time,
  buildDateTime,
}
