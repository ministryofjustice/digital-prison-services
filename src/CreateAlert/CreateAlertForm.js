import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Fieldset from '@govuk-react/fieldset'
import Button from '@govuk-react/button'
import TextArea from '@govuk-react/text-area'
import Select from '@govuk-react/select'
import Link from '@govuk-react/link'

import { connect } from 'react-redux'
import { Form } from 'react-final-form'
import { FORM_ERROR } from 'final-form'
import ErrorSummary from '@govuk-react/error-summary'
import styled from 'styled-components'
import moment from 'moment'

import { linkOnClick } from '../helpers'
import { ButtonContainer, ButtonCancel } from '../Components/Buttons'
import { FieldWithError, onHandleErrorClick, WhenFieldChanges } from '../final-form-govuk-helpers'
import FormDatePicker from '../DatePickers/FormDatePicker'
import { DATE_ONLY_FORMAT_SPEC } from '../dateHelpers'
import FormattedDate from '../DateFormatter'

const DateLinkContainer = styled('div')`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 50%;
`

export const validateThenSubmit = submitHandler => values => {
  const formErrors = []

  if (!values.alertType) {
    formErrors.push({ targetName: 'alertType', text: 'Select a type of alert' })
  }

  if (!values.alertSubType) {
    formErrors.push({ targetName: 'alertSubType', text: 'Select an alert' })
  }

  if (!values.comment) {
    formErrors.push({ targetName: 'comment', text: 'Comment required' })
  }

  if (values.comment && values.comment.length > 1000) {
    formErrors.push({ targetName: 'comment', text: 'Enter a comment using 1000 characters or less' })
  }

  if (!values.effectiveDate) {
    formErrors.push({ targetName: 'effectiveDate', text: 'Select an effective date' })
  }

  if (formErrors.length > 0) return { [FORM_ERROR]: formErrors }

  return submitHandler(values)
}
export class CreateAlertForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      editDate: false,
    }
  }

  render() {
    const {
      alertTypes,
      alertSubTypes,
      cancelHandler,
      handleSubmit,
      errors,
      submitting,
      pristine,
      selectedAlertType,
      effectiveDate,
    } = this.props

    const { editDate } = this.state

    return (
      <form onSubmit={handleSubmit} className="margin-top">
        {Boolean(errors && errors.length) && (
          <ErrorSummary onHandleErrorClick={onHandleErrorClick} heading="There is a problem" errors={errors} />
        )}

        <WhenFieldChanges field="alertType" becomes={selectedAlertType} set="alertSubType" to={undefined} />

        <Fieldset>
          <Fieldset.Legend>Type of alert</Fieldset.Legend>
          <FieldWithError errors={errors} name="alertType" component={Select}>
            <option value="" disabled hidden>
              Select
            </option>
            {alertTypes.map(type => (
              <option key={`alert_type_${type.value}`} value={type.value}>
                {type.description}
              </option>
            ))}
          </FieldWithError>
        </Fieldset>

        <Fieldset>
          <Fieldset.Legend>Alert</Fieldset.Legend>
          <FieldWithError errors={errors} name="alertSubType" component={Select}>
            <option value="" disabled hidden>
              Select
            </option>
            {alertSubTypes.filter(sub => sub.parentValue === selectedAlertType).map(type => (
              <option key={`alert_sub_type_${type.value}`} value={type.value}>
                {type.description}
              </option>
            ))}
          </FieldWithError>
        </Fieldset>

        <Fieldset>
          <Fieldset.Legend>Comment</Fieldset.Legend>
          <FieldWithError name="comment" errors={errors} component={TextArea} />
        </Fieldset>

        <Fieldset>
          <Fieldset.Legend>Effective date</Fieldset.Legend>
          <DateLinkContainer>
            {!editDate && <FormattedDate isoDate={effectiveDate} />}
            {editDate && (
              <FieldWithError
                name="effectiveDate"
                errors={errors}
                component={FormDatePicker}
                placeholder="Select"
                shouldShowDay={date => {
                  const now = moment()
                  const sevenDaysAgo = moment().subtract(7, 'day')
                  const dateToCheck = moment(date, DATE_ONLY_FORMAT_SPEC)

                  return dateToCheck.isSameOrBefore(now, 'date') && dateToCheck.isSameOrAfter(sevenDaysAgo, 'date')
                }}
              />
            )}
            {!editDate && (
              <Link
                data-qa="edit-date-link"
                className="link clickable"
                {...linkOnClick(() =>
                  this.setState(state => ({
                    ...state,
                    editDate: true,
                  }))
                )}
              >
                Change
              </Link>
            )}
          </DateLinkContainer>
        </Fieldset>

        <ButtonContainer className="margin-top-30">
          <Button type="submit" disabled={submitting || pristine} mb={0}>
            Save and continue
          </Button>
          <ButtonCancel mb={0} onClick={cancelHandler} type="button">
            Cancel
          </ButtonCancel>
        </ButtonContainer>
      </form>
    )
  }
}

CreateAlertForm.propTypes = {
  cancelHandler: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  errors: PropTypes.arrayOf(PropTypes.shape({})),
  submitting: PropTypes.bool,
  pristine: PropTypes.bool,
  alertTypes: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ),
  alertSubTypes: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      parentValue: PropTypes.string.isRequired,
    })
  ),
  selectedAlertType: PropTypes.string,
  effectiveDate: PropTypes.string,
}

CreateAlertForm.defaultProps = {
  submitting: false,
  pristine: false,
  alertTypes: [],
  alertSubTypes: [],
  errors: [],
  selectedAlertType: '',
  effectiveDate: '',
}

function CreateAlertFinalForm({ alertTypes, alertSubTypes, cancelHandler, createAlertHandler }) {
  return (
    <Form
      initialValues={{
        effectiveDate: moment(),
      }}
      onSubmit={values => validateThenSubmit(createAlertHandler)(values)}
      render={({ handleSubmit, submitting, pristine, submitError: errors, values }) => (
        <CreateAlertForm
          submitting={submitting}
          pristine={pristine}
          alertTypes={alertTypes}
          alertSubTypes={alertSubTypes}
          handleSubmit={handleSubmit}
          cancelHandler={cancelHandler}
          errors={errors}
          selectedAlertType={values.alertType}
          effectiveDate={values.effectiveDate}
        />
      )}
    />
  )
}

CreateAlertFinalForm.propTypes = {
  cancelHandler: PropTypes.func.isRequired,
  createAlertHandler: PropTypes.func.isRequired,
  alertTypes: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ),
  alertSubTypes: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      parentValue: PropTypes.string.isRequired,
    })
  ),
}

CreateAlertFinalForm.defaultProps = {
  alertTypes: [],
  alertSubTypes: [],
}

export default connect()(CreateAlertFinalForm)
