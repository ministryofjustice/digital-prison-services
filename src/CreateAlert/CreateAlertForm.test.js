import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'

import { CreateAlertForm } from './CreateAlertForm'
import { DATE_ONLY_FORMAT_SPEC } from '../dateHelpers'

const alerts = {
  alertTypes: [{ title: 'alert 1', value: 'alert-type-1' }],
  alertSubTypes: [{ title: 'alert sub 1', value: 'alert-sub-1', parentValue: 'alert-type-1' }],
}

describe('Create alert form', () => {
  const clickChangeDateLink = wrapper => wrapper.find({ 'data-qa': 'edit-date-link' }).simulate('click', {})
  it('should display required fields errors', () => {
    const errors = [
      { targetName: 'alertType', text: 'Please select an alert type' },
      { targetName: 'alertSubType', text: 'Please select an alert' },
      { targetNam: 'comment', text: 'Please enter a comment' },
    ]
    const wrapper = shallow(<CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} errors={errors} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should only contain the default drop down values for alerts and alert sub types', () => {
    const wrapper = shallow(<CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} />)

    const alertTypeOptions = wrapper.find({ name: 'alertType' }).getElement().props.children
    const alertSubTypeOptions = wrapper.find({ name: 'alertSubType' }).getElement().props.children

    expect(alertTypeOptions.slice(1, alertTypeOptions).length).toBe(0)
    expect(
      alertSubTypeOptions.slice(1, alertSubTypeOptions).reduce((acc, current) => acc.concat(current), []).length
    ).toBe(0)
  })

  it('should populate the alert sub types based off the selected alertType', () => {
    const wrapper = shallow(
      <CreateAlertForm
        cancelHandler={() => {}}
        handleSubmit={() => {}}
        selectedAlertType="alert-type-1"
        alertTypes={alerts.alertTypes}
        alertSubTypes={alerts.alertSubTypes}
      />
    )

    const alertSubTypeOptions = wrapper.find({ name: 'alertSubType' }).getElement().props.children

    expect(alertSubTypeOptions[1][0].props.value).toBe('alert-sub-1')
  })

  it('should show FormattedDate by default', () => {
    const wrapper = shallow(
      <CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} effectiveDate="2019-10-10" />
    )

    expect(wrapper.find({ 'data-qa': 'edit-date-link' }).length).toBe(1)
    expect(wrapper.find({ name: 'effectiveDate' }).length).toBe(0)
    expect(wrapper.find('FormattedDate').length).toBe(1)
  })

  it('should hide FormattedDate and show DateInput when the change link is clicked', () => {
    const wrapper = shallow(
      <CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} effectiveDate="2019-10-10" />
    )

    clickChangeDateLink(wrapper)

    expect(wrapper.find({ 'data-qa': 'edit-date-link' }).length).toBe(0)
    expect(wrapper.find({ name: 'effectiveDate' }).length).toBe(1)
    expect(wrapper.find('FormattedDate').length).toBe(0)
  })

  it('should enforce date constraints', () => {
    const wrapper = shallow(
      <CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} effectiveDate="2019-10-10" />
    )
    clickChangeDateLink(wrapper)

    const { shouldShowDay } = wrapper.find({ name: 'effectiveDate' }).getElement().props

    const today = moment().format(DATE_ONLY_FORMAT_SPEC)
    const sevenDaysAgo = moment()
      .subtract(7, 'day')
      .format(DATE_ONLY_FORMAT_SPEC)

    const tomorrow = moment()
      .add(1, 'day')
      .format(DATE_ONLY_FORMAT_SPEC)

    const eightDaysAgo = moment()
      .subtract(8, 'day')
      .format(DATE_ONLY_FORMAT_SPEC)

    expect(shouldShowDay(today)).toBe(true)
    expect(shouldShowDay(sevenDaysAgo)).toBe(true)

    expect(shouldShowDay(tomorrow)).toBe(false)
    expect(shouldShowDay(eightDaysAgo)).toBe(false)
  })

  it('should call cancel handler when cancel button is clicked', () => {
    const cancelHandler = jest.fn()
    const wrapper = shallow(
      <CreateAlertForm cancelHandler={cancelHandler} handleSubmit={() => {}} effectiveDate="2019-10-10" />
    )

    wrapper.find('ButtonCancel').simulate('click', {})

    expect(cancelHandler).toHaveBeenCalled()
  })

  it('should disable submit button when disableSubmit has been passed in', () => {
    const wrapper = shallow(
      <CreateAlertForm cancelHandler={() => {}} handleSubmit={() => {}} effectiveDate="2019-10-10" disableSubmit />
    )

    expect(wrapper.find({ type: 'submit' }).getElement().props.disabled).toBe(true)
  })
})
