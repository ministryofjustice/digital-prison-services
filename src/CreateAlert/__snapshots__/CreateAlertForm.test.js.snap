// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create alert form should display required fields errors 1`] = `
<form
  className="margin-top"
  onSubmit={[Function]}
>
  <ErrorSummary
    errors={
      Array [
        Object {
          "targetName": "alertType",
          "text": "Please select an alert type",
        },
        Object {
          "targetName": "alertSubType",
          "text": "Please select an alert",
        },
        Object {
          "targetNam": "comment",
          "text": "Please enter a comment",
        },
      ]
    }
    heading="There is a problem"
    onHandleErrorClick={[Function]}
  />
  <WhenFieldChanges
    becomes=""
    field="alertType"
    set="alertSubType"
  />
  <styled.fieldset>
    <Legend
      isPageHeading={false}
    >
      Type of alert
    </Legend>
    <FieldWithError
      component={[Function]}
      errors={
        Array [
          Object {
            "targetName": "alertType",
            "text": "Please select an alert type",
          },
          Object {
            "targetName": "alertSubType",
            "text": "Please select an alert",
          },
          Object {
            "targetNam": "comment",
            "text": "Please enter a comment",
          },
        ]
      }
      name="alertType"
    >
      <option
        disabled={true}
        hidden={true}
        value=""
      >
        Select
      </option>
    </FieldWithError>
  </styled.fieldset>
  <styled.fieldset>
    <Legend
      isPageHeading={false}
    >
      Alert
    </Legend>
    <FieldWithError
      component={[Function]}
      errors={
        Array [
          Object {
            "targetName": "alertType",
            "text": "Please select an alert type",
          },
          Object {
            "targetName": "alertSubType",
            "text": "Please select an alert",
          },
          Object {
            "targetNam": "comment",
            "text": "Please enter a comment",
          },
        ]
      }
      name="alertSubType"
    >
      <option
        disabled={true}
        hidden={true}
        value=""
      >
        Select
      </option>
    </FieldWithError>
  </styled.fieldset>
  <styled.fieldset>
    <Legend
      isPageHeading={false}
    >
      Comment
    </Legend>
    <FieldWithError
      component={[Function]}
      errors={
        Array [
          Object {
            "targetName": "alertType",
            "text": "Please select an alert type",
          },
          Object {
            "targetName": "alertSubType",
            "text": "Please select an alert",
          },
          Object {
            "targetNam": "comment",
            "text": "Please enter a comment",
          },
        ]
      }
      name="comment"
    />
  </styled.fieldset>
  <styled.fieldset>
    <Legend
      isPageHeading={false}
    >
      Effective date
    </Legend>
    <styled.div>
      <FormattedDate
        isoDate=""
      />
      <styled.a
        className="link clickable"
        data-qa="edit-date-link"
        muted={false}
        noVisitedState={false}
        onClick={[Function]}
        onKeyDown={[Function]}
        role="link"
        tabIndex={0}
        textColour={false}
      >
        Change
      </styled.a>
    </styled.div>
  </styled.fieldset>
  <styled.div
    className="margin-top-30"
  >
    <ForwardRef
      disabled={false}
      mb={0}
      start={false}
      type="submit"
    >
      Save and continue
    </ForwardRef>
    <ButtonCancel
      mb={0}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </ButtonCancel>
  </styled.div>
</form>
`;
